### PR TITLE
Modifications to allow papas to work with lxplus and Gaudi

### PR DESCRIPTION
**Pull request by alicerobson**
This includes several small modifications made to allow papas to run under lxplus/gaudi. Including:-
Make particles const for simulator (see also issue #32) which affects the Path.
Issue with subclusters (I think I could later on get rid of subclusters and just have ids which would be safer, see issue #31 )
Shared_ptrs rationalised and updated to C++14
Add in missing virtual destructors to base classes (for inheritance)
Other minor tidying etc
There are some additional destructors added which allow a temp workaround to the seg fault. These will hopefully be removed shortly.
There is also a rationalisation of the history as used in papasmanager (issue #8)

### CODE REVIEW

File: init.sh [hunk 1]
```diff
@@ -2,17 +2,18 @@ platform='unknown'
...
     echo "FCCPAPASCPP is unset, setting to $PWD/install";
     export FCCPAPASCPP=$PWD/install
     export PATH=$FCCPAPASCPP/bin:$PATH    
-#fi
+fi
 
 if [[ "$unamestr" == 'Linux' ]]; then
     platform='Linux'
     export LD_LIBRARY_PATH=$FCCPAPASCPP/lib:$LD_LIBRARY_PATH
+    source /cvmfs/fcc.cern.ch/sw/0.8/init_fcc_stack.sh $1
```
Comment (jlingema) [inline init.sh]: Hi Alice,

Can you update this to 0.8.1? That should also fix the problem that made the test fail.

Cheers,
Joschka

**Comment by alicerobson** [comment]
Thanks Joschka,

I am on the case!
A

**Comment by alicerobson** [comment]
Joschka

Would you be able to check the Jenkins part of this. I have put the latest setup into the papas init.sh file but the Jenkins report  still seems to be saying it is setting up 0.8. I seem to remember that the setup was not always in the init.sh file and thus was in Jenkins, so I am wondering if a change (removal from)  the Jenkins setup is required. It seems fine when I compile.

Many Thanks
Alice

**Comment by jlingema** [comment]
Hi Alice, indeed I removed it from the Jenkins setup.

retest this please.

File: papas/datatypes/Cluster.h [hunk 2]
```diff
@@ -26,7 +26,7 @@ class Cluster {
    @param[in]  id identifier type of cluster eg kEcalCluster or kHcalCluster
    @param[in]  subtype single char describing type of cluster eg s = smeared, t= true, m = merged
     */
-  Cluster(double energy, const TVector3& position, double size_m, unsigned int index, IdCoder::ItemType id,
+  Cluster(double energy, const TVector3& position, double size_m, uint32_t index, IdCoder::ItemType id,
```
Comment (cbernet) [inline papas/datatypes/Cluster.h]: why this change?

File: papas/datatypes/Event.h [hunk 3]
```diff
@@ -58,7 +58,7 @@ namespace papas {
 class Event {
 public:
   /// @brief  Constructor
-  Event(std::shared_ptr<Nodes> hist = std::make_shared<Nodes>(Nodes()));
+  Event(Nodes& hist);
```
Comment (cbernet) [inline papas/datatypes/Event.h]: What is the new type of Nodes?

File: papas/datatypes/src/Cluster.cpp [hunk 4]
```diff
@@ -8,22 +8,22 @@ namespace papas {
 
 double Cluster::s_maxEnergy = 0;
 
-Cluster::Cluster(double energy, const TVector3& position, double size_m, unsigned int index, IdCoder::ItemType type,
+Cluster::Cluster(double energy, const TVector3& position, double size_m, uint32_t index, IdCoder::ItemType type,
                  char subtype)
-    : m_id(IdCoder::makeId(index, type, subtype, fmax(0, energy))), m_position(position), m_subClusters({this}) {
+    : m_id(IdCoder::makeId(index, type, subtype, fmax(0, energy))), m_position(position), m_subClusters({}) {
```
Comment (cbernet) [inline papas/datatypes/src/Cluster.cpp]: Why initializing m_subClusters? it's probably done by default if you remove it.

File: papas/simulation/Propagator.h [hunk 5]
```diff
@@ -17,19 +17,26 @@ class Propagator {
...
+  virtual ~Propagator() = default;
+
   /**
    Propagate particle to the selected cylinder and store the point where the particle crossed the cylinder
    @param[in] ptc particle that is to be propagated
    @param[in] cyl cylinder to which the particle is to be propagated.
    @param[in] field magnitude of magnetic field (used only for charged particles, defaults to zero if not set)
    */
-  virtual void propagateOne(Particle& ptc, const SurfaceCylinder& cyl) const = 0;
+  virtual void propagateOne(const Particle& ptc, const SurfaceCylinder& cyl) const = 0;
```
Comment (cbernet) [inline papas/simulation/Propagator.h]: One can the path of ptc be modified if the particle is const?

File: papas/simulation/HelixPropagator.h [hunk 6]
```diff
@@ -21,7 +21,13 @@ class HelixPropagator : public Propagator {
    @param[in] cyl cylinder to which the particle is to be propagated.
    @param[in] field magnitude of magnetic field
    */
-  virtual void propagateOne(Particle& ptc, const SurfaceCylinder& cyl) const override;
+  virtual void propagateOne(const Particle& ptc, const SurfaceCylinder& cyl) const override;
+
+  /** Sets the particle path to a helix
+   @param[in] ptc particle that is to be propagated
+   */
+  virtual void setPath(Particle& ptc) const override;
```
Comment (cbernet) [inline papas/simulation/HelixPropagator.h]: Where is setPath called?

File: papas/simulation/Propagator.h (hunk 5 above)
Comment (alicerobson) [inline papas/simulation/Propagator.h]: Yes, it can (the particle contains the shared_pointer to the path and the shared_pointer is const).
There is more discussion of this in Design of Particle and Path  Issue #32:-
https://github.com/HEP-FCC/papas/issues/32

File: papas/simulation/HelixPropagator.h (hunk 6 above)
Comment (alicerobson) [inline papas/simulation/HelixPropagator.h]: Its used in PFReconstructor reconstructCluster and reconstructTrack
https://github.com/alicerobson/papas/blob/lxplus/papas/reconstruction/src/PFReconstructor.cpp#L293

File: papas/datatypes/src/Cluster.cpp (hunk 4 above)
Comment (alicerobson) [inline papas/datatypes/src/Cluster.cpp]: Yes, thanks. Done.

File: papas/datatypes/Event.h (hunk 3 above)
Comment (alicerobson) [inline papas/datatypes/Event.h]: Sorry, I don't understand your question. Is this what you are after:-
typedef DAG::Node<Identifier> PFNode;
typedef std::map<Identifier, PFNode> Nodes;

File: papas/datatypes/Cluster.h (hunk 2 above)
Comment (alicerobson) [inline papas/datatypes/Cluster.h]: The index is used as a component of the Id using bitshifts etc. I wondered if the bug on lxplus, could have anything to do with the Id (eg writing off the end somehow) so tried to be extra careful. I believe that unsigned int allows for different size implementation on different systems. The Id is fully specified as uint64_t and so I decided it would be safest to also be fully explicit as to the size of the index part of this so that it is guaranteed to be the same on all operating systems.
